common/kvstore/src/main/java/org/apache/spark/util/kvstore/LevelDBIterator.java

```diff
@@ -182,23 +193,36 @@ public boolean skip(long n) {
 
   @Override
   public synchronized void close() throws IOException {
-    db.notifyIteratorClosed(this);
+    db.notifyIteratorClosed(it);
     if (!closed) {
-      it.close();
-      closed = true;
-      next = null;
+      try {
+        it.close();
+      } catch (UncheckedIOException uncheckedIOException) {
+        throw uncheckedIOException.getCause();
+      } finally {
+        closed = true;
+        next = null;
+        cancelResourceClean();
+      }
     }
   }
 
   /**
-   * Because it's tricky to expose closeable iterators through many internal APIs, especially
-   * when Scala wrappers are used, this makes sure that, hopefully, the JNI resources held by
-   * the iterator will eventually be released.
+   * Prevent ResourceCleaner from actually releasing resources after close it.
    */
-  @SuppressWarnings("deprecation")
-  @Override
-  protected void finalize() throws Throwable {
-    db.closeIterator(this);
+  private void cancelResourceClean() {
+    this.resourceCleaner.setStartedToFalse();
+    this.cleanable.clean();
```

Comment:
We have already notified and closed the iterator before this method is invoked - and we are forcing a clean again ?

Reply:
This was a suggestion I made earlier. Line 214 sets the `started` status of `ResourceCleaner` to false, while line 215 simply triggers ResourceCleaner to manually execute and be removed from the Cleaner's pending cleanup queue as soon as possible. In fact, since the `started` status is already `false`, it won't perform any operations. Of course, we could also just set the status here and then wait for the `Cleaner` to execute the registered ResourceCleaner on its own ：）

Reply:
Yes, but the clean is a noop right (since started has been force set to `false`) ? Is there a reason to call it ?

Reply:
In other words, what I am trying to understand is (should have phrased it better :) ) what is the overhead/issues with cleaner ending up invoking this as its normal case.

(Agree with your rationale btw - makes sense to evict it from queue).

Reply:
Yes, it's a no-op. Even if we don't call it, the Cleaner will invoke the clean method on objects that have been garbage collected. Personally, I prefer to have the Cleaner execute more effective resource reclamation when it calls ResourceCleaner, rather than such no-ops.

Reply:
My concern is: if we don't manually call the `clean` method here, when the `Cleaner` actively performs GC, it will accumulate a large number of 'no-op' calls, which prevents tasks that really need to be recycled from being executed as soon as possible.

Reply:
However, I indeed don't have valid test data yet to support my viewpoint.

Reply:
It makes sense, let us keep it as is - I wanted to make sure I was not missing something !